libs/libarchfpga/src/physical_types.h

```diff
@@ -874,6 +874,19 @@ struct t_pb_graph_node {
 
 	int placement_index;
 
+	/* Contains a collection of mode indices that cannot be used as they produce conflicts during VPR packing stage
+	 *
+	 * Illegal modes do arise when children of a graph_node do have inconsistent `edge_modes` with respect to
+	 * the parent_pb.
+	 * Example: Edges that connect LUTs A, B and C to the parent pb_graph_node refer to the correct parent's mode which is set to "LUTs",
+	 *          but edges of LUT D have the mode of edge corresponding to a wrong parent's pb_graph_node mode, namely "DRAMs".
```

Comment:
Thanks for the comment; this gives a  good big picture. A few more tweaks:

1. Should say "LUTRAM" instead of "DRAMs".  (DRAMs are off chip.  BRAMs are on-chip, but LUTRAM is the name usually used for a logic block that has converted to BRAM on demand).
2. Is the mode flagged as illegal a randomly chosen one of the conflicting modes?  For example, here we could mark "LUTs" illegal or "DRAM/LUTRAM" as illegal.  Is the choice of marking LUTs as a mode conflict arbitrary, or is there some heuristic?  Worth mentioning what is done.
3. (small nit): fix spelling of inconsistent (not incosistent).

Reply:
Ok, thanks for the feedback. I'll update the comments. 

Anyway the conflicting mode marked as illegal is the most restrictive one. 

In the example, `LUTs` mode is the one that is chosen for LUTs A, B and C. LUT D instead couldn't use `LUTs` as there is no possible path using this mode. Therefore, the LUT D edge that connects it to the parent `pb_graph_node` refers to another mode (`LUTRAM` in our example). `LUTs` mode for the parent `pb_graph_node` is marked as illegal as LUT A, B and C can still find a route using the `LUTRAM` mode of the parent `pb_graph_node`.
If no mode is found that satisfies the checks it means that there is something wrong with the architecture definition.

This of course is just an example, but I hope it clarifies the situation.

Reply:
Corrected the comments.

Reply:
Thanks!  It looks good and I'm pretty sure I understand the overall picture now.

Reply:
Perfect, I am glad to hear that, if there is something that is not yet 100% clear do not hesitate to ask. @litghost and @mithro could provide further insights